### DISCUSSION TITLE
Add search statement or filter beatmap on search bar based on max pp

### DISCUSSION THREAD

**Discussion by huyenden** [ppy/osu#24896] [Ideas] [answered]
The best way for players to improve their skills is to play beatmaps that are higher than the maximum maxpp they can reach and not exceed them too much.

When playing multiple beatmaps with the same number of stars, I realized that the difficulty of completing 100% of the beatmap is different because OD affects maxpp a lot, leading to one map being more difficult than the other.

Then I realized, with 2 beatmaps with equal max pp, map A with the lower number of stars will be the Accuracy map and map B with the higher number of stars will be the Aim map.

Practical example that with the same maxpp:

100pp:
**Accuracy:** https://osu.ppy.sh/beatmapsets/953265#osu/1990363 **[3.53*]**
**Aim:** https://osu.ppy.sh/beatmapsets/117723#osu/323960 **[4.39*]**

200pp:
**Accuracy:** https://osu.ppy.sh/beatmapsets/1424727#osu/2933980 **[4.67*]**
**Aim:** https://osu.ppy.sh/beatmapsets/641201#osu/1613453 **[5.58*]**

From then on I became less dependent on searching for beatmaps based on the number of stars. Because they are unstable to find the beatmap that suits you. Because it's easier to play a 4* aim map with 100pp well than a 4* accuracy map with 200pp, but players don't know clearly whether that 4* map is an aim or accuracy map if they just look at the stars.

Being disoriented and wasting a lot of time finding the right beatmap to improve my pp score based on the 100k ranked map gives me a headache.

Sometimes I'm stuck in the middle of life and tired of searching for beatmaps based on max pp from discord bots. Or see information about what players can achieve on the beatmap rankings. Just to find a beatmap that matches my current maxpp

I don't know if there is any software that supports this search, but if there was a built-in search command in osu!lazer, and also a beatmap download area, that would be a lot more convenient. I'm very serious about this suggestion to further improve osu!lazer.

![Ảnh chụp màn hình 2023-09-24 032251](https://github.com/ppy/osu/assets/128424762/cb3d6817-7256-4853-82de-48383c237de0)

**Comment by Rekunan** [comment]
There are plans to have star rating be directly correlated with pp, so OD (length, and other factors) will also affect SR in the future.

And to help with your search in aim maps, there are some talks about having skill splitting, [similar to osu!tp](https://twitter.com/minusgn/status/1569104303901790210), sometime in the far future, with one of the skills being aim.

**Reply by huyenden** [reply]
> There are plans to have star rating be directly correlated with pp, so OD (length, and other factors) will also affect SR in the future.
> 
> And to help with your search in aim maps, there are some talks about having skill splitting, [similar to osu!tp](https://twitter.com/minusgn/status/1569104303901790210), sometime in the far future, with one of the skills being aim.

Currently the osutp.net website is not working, I also don't know if the website displays the maxpp of ranked beatmaps or not. If -GN ever mentioned it, there is still no current solution

**Reply by Rekunan** [reply]
> I also don't know if the website displays the maxpp of ranked beatmaps or not

It doesn't, and my comment was meant as an alternative of displaying the maxpp, but I'm realizing it's really convenient to also display that regardless.

**Comment by huyenden** [comment]
Reply pls 🥺

**Reply by Rekunan** [reply]
They're probably too busy with making lazer ranked to respond to something like this rn.

**Comment by bdach** [comment, marked as answer]
The big elephant in the room is, what about mods? They'll influence max pp.

You could do what star rating filter does right now and only consider max _nomod_ pp. But I don't know that players don't want mods included in that filter either, possibly, if we're spitballing potential improvements.

**Reply by huyenden** [reply]
Agree considering adding mods (similar to dt,hd,hr or Difficulty Adjustment OD11) will change the max pp of the map and make search commands complicated, making it difficult to Search maximum pp in Bancho database and internal device.

**Reply by huyenden** [reply]
In any case, I cannot selfishly ask too much of the developers. Please feel free to work, you all are still dedicating yourself to the community 👍

**Comment by SevenChords** [comment]
I'd like to see this as well.
Since the game client already includes a way to see the maximum available PP in the score screen, it should be possible to add a display of the maximum available PP to the left info section during score select and a search filter (e.g. `pp>100` to only show beatmaps with more than 100pp achievable without mods (or, if possible: with currently selected mods)).


I can see a couple of use cases for this:

1. Filter out all songs that you cannot gain overall pp from (e.g. if the lowest weighted score from your top 100 plays is 90pp, filtering `pp>90` would only show songs where it is possible to achieve a score that lands in your top 100 ranked plays).
2. If you want to aim for certain pp amounts in resulting scores to only have those visible (e.g. you want to achieve a 200pp score, filter `pp>200` to only display songs where that would be possible).


If it's possible to access the data of the available PP that is shown on the results screen during song select, it shouldn't be too hard to implement i think.

**Reply by bdach** [reply]
again, maximum available pp _with what mods_?

it's a bit different to do that on results screen, with one score, and known mods. it's another to do it for all maps of however many you might have locally, and potentially arbitrary mod combinations.

**Reply by huyenden** [reply]
If it was 1 year ago I think my proposal would have been fine, but after 1 year I think I agree with bdach more.

In osu!stable, players can calculate and preview the maxpp of a beatmap via the Tillerino bot, or a very convenient overlay like [gosumemory](https://github.com/l3lackShark/gosumemory) or [tosu](https://github.com/KotRikD/tosu).

McOsu also has live maxpp display feature.

The problem here is suppose the player has 100k ranked map. Then if there is a maxpp calculation algorithm command in the search bar. Then every time switching between mods like hr, hd, dt. Will cause terrible lag because of calculating 100k beatmap. Importing maps is similar.

**Reply by bdach** [reply]
we can probably precalc this data and store this to database, but it becomes an issue (every pp/sr rebalance means everything has to be recalculated etc.)

**Reply by huyenden** [reply]
@bdach [#29553](https://github.com/ppy/osu/pull/29553) now it works fine thank you. now i guess playing without missing ranked maps. looks more convenient for completionists now

**Reply by huyenden** [reply]
> we can probably precalc this data and store this to database, but it becomes an issue (every pp/sr rebalance means everything has to be recalculated etc.)

For me, if there was actually a button to calculate the maxpp available for a specific mod that the player has selected in the mod settings, like NM or HDDT and save it to the client.realm database to search for pp on the search bar, I think there would be no problem for the player if they had to wait or have a rework of sr/pp. The player would just have that button and everything would be fine. To search beatmaps by pp on the search bar is similar to my topic last year.
![IMG_1506](https://github.com/user-attachments/assets/5cc5822f-87c4-429d-9194-15b701d5ccb3)

**Reply by SevenChords** [reply]
> again, maximum available pp _with what mods_?
> 
> it's a bit different to do that on results screen, with one score, and known mods. it's another to do it for all maps of however many you might have locally, and potentially arbitrary mod combinations.

Since NONE of the current search options take mods into account and all filters so far are only for nomod, this would be an easy choice to do it for nomod, and since most mods give a relative boost to pp players could estimate that boost themselves, just as the boost to star rating, bpm etc.
